rust/catalyst-types/src/id_uri/role_index.rs

```diff
@@ -12,54 +12,76 @@ use thiserror::Error;
 /// Role Index parsing error
 #[derive(Display, Error, Debug)]
 #[allow(clippy::module_name_repetitions)]
-pub enum RoleIndexError {
+pub enum RoleIdError {
+    /// Invalid parse
+    Parse(#[from] ParseIntError),
+
     /// Invalid Role Index
-    InvalidRole(#[from] ParseIntError),
+    InvalidRole(u8),
 }
 
 /// Project Catalyst User Role Index.
 ///
 /// <https://github.com/input-output-hk/catalyst-CIPs/blob/x509-catalyst-role-definitions/CIP-XXXX/README.md>
 #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct RoleIndex(u16);
-
-impl RoleIndex {
-    /// Delegated Representative
-    pub const DREP: RoleIndex = RoleIndex(1);
-    /// Proposer
-    pub const PROPOSER: RoleIndex = RoleIndex(3);
-    /// Default Role 0
-    pub const ROLE_0: RoleIndex = RoleIndex(0);
+#[repr(u8)]
+#[non_exhaustive]
+#[derive(strum::FromRepr)]
+pub enum RoleId {
+    /// Primary required role use for voting and commenting.
+    Role0 = 0,
+    /// Delegated representative (dRep) that vote on behalf of delegators.
+    DelegatedRepresentative = 1,
+    /// Voter role that delegates voting power to a chosen representative (dRep).
+    VoterDelegation = 2,
+    /// Proposer that enabling creation, collaboration, and submission of proposals.
+    Proposer = 3,
+}
 
-    /// Is the `RoleIndex` the default value (Role 0)
+impl RoleId {
+    /// Is the `RoleId` the default value [`Self::Role0`]
     #[must_use]
     pub fn is_default(self) -> bool {
-        self == Self::ROLE_0
+        self == Self::Role0
     }
 }
 
-impl Default for RoleIndex {
+impl Default for RoleId {
     fn default() -> Self {
-        Self::ROLE_0
+        Self::Role0
     }
 }
 
-impl From<u16> for RoleIndex {
-    fn from(value: u16) -> Self {
-        Self(value)
+impl TryFrom<u8> for RoleId {
+    type Error = RoleIdError;
+
+    fn try_from(value: u8) -> Result<Self, RoleIdError> {
+        Self::from_repr(value).ok_or(RoleIdError::InvalidRole(value))
     }
 }
 
-impl FromStr for RoleIndex {
-    type Err = RoleIndexError;
+impl FromStr for RoleId {
+    type Err = RoleIdError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(Self(s.parse::<u16>()?))
+        let value: u8 = s.parse()?;
+
+        Self::try_from(value)
     }
 }
 
-impl Display for RoleIndex {
+impl Display for RoleId {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
-        write!(f, "{}", self.0)
+        write!(f, "{}", *self as u8)
+    }
+}
+
+impl<'a, C> minicbor::Decode<'a, C> for RoleId {
```

Comment:
We should have a unit test to check various cbor encodings deserialize properly.
We should also implement cbor encode for completeness.